src/util/triplea/MapXMLCreator/CanalDefinitionsPanel.java

```diff
@@ -0,0 +1,442 @@
+              "Input Error", JOptionPane.ERROR_MESSAGE);
+          currentCanalName = JOptionPane.showInputDialog(null,
+              "Which canal should be selected for territory '" + newTerrName + "?", currentCanalName);
+        }
+      } else if (currentCanalName != null) {
+        final Tuple<SortedSet<String>, SortedSet<String>> canalTerrs =
+            MapXMLHelper.canalDefinitions.get(currentCanalName);
+        selectedWaterTerritories = canalTerrs.getFirst();
+        selectedLandTerritories = canalTerrs.getSecond();
+        SwingUtilities.invokeLater(new Runnable() {
+          public void run() {
+            imagePanel.repaint();
+          }
+        });
+        return;
+      }
+      if (currentCanalName == null)
+        return;
+    }
+
+    if (currentCanalName != null) {
+      Tuple<SortedSet<String>, SortedSet<String>> canalTerrs = MapXMLHelper.canalDefinitions.get(currentCanalName);
+      if (canalTerrs == null) {
+        canalTerrs = Tuple.of(new TreeSet<String>(), new TreeSet<String>());
+        MapXMLHelper.putCanalDefinitions(currentCanalName, canalTerrs);
+      }
+
+      selectedWaterTerritories = canalTerrs.getFirst();
+      selectedLandTerritories = canalTerrs.getSecond();
+
+      final SortedSet<String> selectedTerrsSameType;
+      if (newTerrIsWater)
+        selectedTerrsSameType = selectedWaterTerritories;
+      else
+        selectedTerrsSameType = selectedLandTerritories;
+
+      if (selectedTerrsSameType.size() > 0) {
+        final SortedSet<String> commonNeighborsDiffType =
+            getCommonNeighborsOfType(selectedTerrsSameType, !newTerrIsWater);
+        commonNeighborsDiffType.retainAll(newTerrNeighborsDiffType);
+
+        if (commonNeighborsDiffType.size() < 2) {
+          JOptionPane.showMessageDialog(null, "The selected " + (newTerrIsWater ? "water" : "land")
+              + " territory is connected to less than 2 common " + (!newTerrIsWater ? "water" : "land")
+              + " territories with the other " + (newTerrIsWater ? "water" : "land") + " territor"
+              + (selectedTerrsSameType.size() == 1 ? "y" : "ies")
+              + "!\rRight click to deselect current canal '" + currentCanalName + "'.", "Input Error",
+              JOptionPane.ERROR_MESSAGE);
+          return;
+        }
+      }
+
+      if (selectedTerrsSameType.contains(newTerrName))
+        selectedTerrsSameType.remove(newTerrName);
+      else
+        selectedTerrsSameType.add(newTerrName);
+
+    }
+
+    SwingUtilities.invokeLater(new Runnable() {
+      public void run() {
+        imagePanel.repaint();
+      }
+    });
+  }
+
+  protected void clearSelection() {
+    currentCanalName = null;
+    selectedLandTerritories = new TreeSet<String>();
+    selectedWaterTerritories = new TreeSet<String>();
+  }
+
+  protected LinkedHashSet<String> getNeighborsByType(final String newTerrName, final boolean waterNeighbors) {

```

Comment:
Avoid methods with boolean parameters. Favor creating two different methods to capture the two different cases. Alternatively when that is not an improvement, favor using an enum instead of a boolean.